File: cpp/example/trigger_message_example.cc
#include <drift_protocol/common/drift_package.pb.h>
#include <drift_protocol/trigger_service/interval_trigger_message.pb.h>
#include <google/protobuf/util/time_util.h>

using drift::proto::trigger_service::IntervalTriggerMessage;
using drift::proto::common::DriftPackage;
using drift::proto::common::StatusCode;

using google::protobuf::util::TimeUtil;
using google::protobuf::Any;

int main() {
    const auto pb_time = TimeUtil::GetCurrentTime();
    std::string message;

    {
        // Create a package and serialize it
        DriftPackage original;
        original.set_id(TimeUtil::TimestampToMilliseconds(pb_time)); // UNIX timestamp in ms
        original.set_status(StatusCode::GOOD);
        original.mutable_source_timestamp()->CopyFrom(pb_time);
        original.mutable_publish_timestamp()->CopyFrom(pb_time);

        // Trigger message for 1 second interval
        IntervalTriggerMessage trigger;
        trigger.mutable_start_timestamp()->CopyFrom(pb_time - TimeUtil::SecondsToDuration(1));
        trigger.mutable_stop_timestamp()->CopyFrom(pb_time);

        // Serialize package to message
        original.add_data()->PackFrom(trigger);
        original.SerializePartialToString(&message);
    }

    // Parse the package with the trigger
    DriftPackage new_package;
    new_package.ParseFromString(message);
    std::cout << "Package ID=" << new_package.id() << std::endl;

    IntervalTriggerMessage trigger;
    new_package.data(0).UnpackTo(&trigger);
    std::cout << "Trigger interval from " << TimeUtil::ToString(trigger.start_timestamp()) << " to "
              << TimeUtil::ToString(trigger.stop_timestamp()) << std::endl;
}

File: python/examples/trigger_message_example.py
from google.protobuf.any_pb2 import Any
from google.protobuf.timestamp_pb2 import Timestamp

from drift_protocol.common import DriftPackage, StatusCode
from drift_protocol.trigger_service import IntervalTriggerMessage

if __name__ == "__main__":
    pb_time = Timestamp()
    pb_time.GetCurrentTime()

    # Create a package and serialize it
    original = DriftPackage()
    original.id = pb_time.ToMilliseconds()
    original.source_timestamp.CopyFrom(pb_time)
    original.publish_timestamp.CopyFrom(pb_time)
    original.status = StatusCode.GOOD

    # Trigger message for 1 second interval
    trigger = IntervalTriggerMessage()
    trigger.start_timestamp.FromMilliseconds(pb_time.ToMilliseconds() - 1_000_000)
    trigger.stop_timestamp.CopyFrom(pb_time)

    msg = Any()
    msg.Pack(trigger)
    original.data.append(msg)

    # Serialize package to message
    message = original.SerializeToString()

    # Parse the package
    new_pacakge = DriftPackage()
    new_pacakge.ParseFromString(message)
    print(f"Package ID={new_pacakge.id}")

    trigger = IntervalTriggerMessage()
    new_pacakge.data[0].Unpack(trigger)
    print(f"Trigger interval from from {trigger.start_timestamp.ToDatetime()} to {trigger.stop_timestamp.ToDatetime()}")
